[MRESOLVER-228] skip & reconcile

caiwei-ebay opened the pull request.

michael-o reviewed (changes requested):
Those verbose boolean properties are wrong. They are an antipattern to log levels. That should go to `DEBUG`.

I will run the code as-is against our Core ITs first.

michael-o commented:
It does not cause at least any failures in Core ITs...

michael-o commented on maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/ConflictWinnerFinder.java:
```diff
@@ -0,0 +1,255 @@
+ */
+
+import com.google.common.collect.LinkedHashMultimap;
+import com.google.common.collect.Multimap;
```
We don't use Guava directly in Resolver and Core. It is a transtive dep of Guice we cannot get rid of. Don't use it please. @cstamas Do you agree here?

cstamas replied:
Yup, in Java8 it is easy to by out Multimap using getOrDefault etc

michael-o commented on maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/ConflictWinnerFinder.java:
```diff
@@ -0,0 +1,255 @@
+     * @return
+     * @throws DependencyCollectionException
+     */
+    private Multimap<String, Conflict> resolveConflictWinners( RepositorySystemSession session, CollectResult res )
```
res => result

michael-o commented on maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/ConflictWinnerFinder.java:
```diff
@@ -0,0 +1,255 @@
+     *
+     * @return
+     */
+    public Collection<Conflict> getVersionConflicts( RepositorySystemSession session, CollectResult res )
```
res => result

michael-o commented on maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/ConflictWinnerFinder.java:
```diff
@@ -0,0 +1,255 @@
+        for ( String key : conflictMap.keySet() )
+        {
+            Collection<Conflict> col = conflictMap.get( key );
+            if ( col != null && col.size() > 1 )
```
Explicitly more than one element? If, yes add a comment. Others might assume a typo here.

michael-o commented on maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/DefaultDependencyCollector.java:
```diff
@@ -80,6 +80,17 @@
+     * The key in the repository session's {@link org.eclipse.aether.RepositorySystemSession#getConfigProperties()
+     * configuration properties} used to store a {@link Boolean} flag controlling the resolver's skip & reconcile mode.
+     */
+    public static final String CONFIG_PROP_RESOLVER_MODE = "resolver.mode";
```
A boolean flag contradicts the property name. `mode` implies an enum of options. See other modes: https://maven.apache.org/resolver-archives/resolver-LATEST/configuration.html

michael-o commented on maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/DefaultDependencyCollector.java:
```diff
@@ -80,6 +80,17 @@
+     * The key in the repository session's {@link org.eclipse.aether.RepositorySystemSession#getConfigProperties()
+     * configuration properties} used to store a {@link Boolean} flag controlling the resolver's verbose mode.
+     */
+    public static final String CONFIG_PROP_RESOLVER_VERBOSE_MODE = "resolver.verbose";
```
As mentioned, this should go for proper logging.

cstamas commented:
Kudos, cool work!

cstamas commented:
@caiwei-ebay @michael-o took the liberty to create a draft as I had quite a few remarks https://github.com/apache/maven-resolver/pull/137

Most important IMHO is to keep the logic flow "clean" -- so either use reconciler or not use, passing in boolean to "enable" it, and then have all method to some no-op is IMHO wrong. Just create reconciler when used, otherwise do not.

caiwei-ebay commented:
@michael-o @cstamas 

Integrated the changes from #137 to this PR and do a bit refactoring.
Please help review the PR.

Thanks!

caiwei-ebay commented:
@michael-o @cstamas 
Is there anything I can help further?
Please share your thoughts on this PR. Would like to refine the PR as you suggested.

michael-o commented:
> @michael-o @cstamas Is there anything I can help further? Please share your thoughts on this PR. Would like to refine the PR as you suggested.

I am still reviewing another open PR, then I can come to your one. Don't expect any huge traction in the next 10 days.

caiwei-ebay commented:
Basically the strategy of this approach is:
- Skip more nodes
- Reconcile least nodes

There is one more case that the node can be skipped reconciling: the node's parents includes any conflict losers.
In such case, maven won't pick up such node at all, so it is safe to skip reconciling.
Also renamed a few variables, added a few comments to make it easy to understand.

Here is the commit:
https://github.com/apache/maven-resolver/pull/136/commits/78e5e98b6430ba71c982cbc9bce6f603cc2cb286

We have recertified this code changes by running large numbers of applications in our company by comparing dependency:tree and dependency:list result.

Now most of the applications skipped quite a few nodes and just need to reconcile 0 nodes.
And if there is version conflicts, this approach just needs to reconcile least nodes.

caiwei-ebay commented:
@michael-o @cstamas 

Squashed the commits. Simplified the algorithm a lot.

 **Skip:** 
 
Skip resolving the node if a node with the same GAV and lower (or equal) depth has been calculated before.
Sample:

- A -> B -> C (C calculated and cached with depth==3)
-  A -> E -> F -> C (C will be skipped and children is not set, record this C has been skipped by above C node)
-  A -> C (C re-calculated and cached with depth==2)
- A -> K -> C (C will be skipped, record this C has been skipped by the C node with path A -> C)
	
With skip step, there is no result change with "mvn dependency:tree -Dverbose", and as there would be dependency conflicts,
so "mvn dependency:tree" or "mvn dependency:list" result might be impacted, thus we need to reconcile.


**Reconcile:**

-  A  ->   B   ->  C 3.0 ->  D  ->  Z        (D of C 3.0 calculated and cached with depth==4)
- A  ->   E   ->    F   ->  G  ->  D  -> Z  (D will be skipped and children is not set because depth 5 is deeper, record this D has been skipped by above D node)
- A  -> C 2.0 ->    H                       (C 2.0 comes to resolve and maven will pick C 2.0 as depth is lower. The D of C3.0 in step 1 is no longer valid, need reconcile the skipped D node in step 2)
	
Here is the flow: 

- Get the cloned root node A by deep cloning the original root node, the original root node has already been applied above skip strategy. 

```
A  ->   B   ->  C 3.0 ->  D  ->  Z
A  ->   E   ->    F   ->  G  ->  D   (D has no children as skipped)
A  -> C 2.0 ->    H
```


- Call Session's transformer to transform the cloned root node A with verbose mode, note transformer will set children as empty for conflict losers (ConflictResolver.removeLosers).

```
A  ->   B   ->  C 3.0                (C 3.0 has no children as chopped by transformer)
A  ->   E   ->    F   ->  G  ->  D   (D has no children as skipped)
A  -> C 2.0 ->    H
```


- Use ReconclingClonedGraphVisitor to iterate the transformed cloned root node. When visitor comes to a node without children, it can either a skipped node with no children set or chopped by transformer in step 2)

   We just need to reconcile all skipped nodes (the node has no children and children of original node before transform is also empty).
   
   `D (parent path A -> E -> F -> G) needs to be reconciled`
   
   The visitor also record the dependency conflicts according to the ConflictResolver.NODE_DATA_WINNER property.
   
   `C 2.0 conflicts with C 3.0`
   

-  Remove cached items from nodesWithDepth if the parent paths of current node includes any above conflict losers.

 ```
   remove D (children Z， parent path A -> B -> C3.0) from nodesWithDepth
   remove Z (parent path A -> B -> C3.0 -> D) from nodesWithDepth
```

-  Reconcile the nodes (recalculate their children) found in step 3.

   
   `recalculate D (parent path A -> E -> F -> G, skipped by A -> B -> C3.0 -> D)`

cstamas commented:
pls incorporate latest master changes (whether rebase+squash+force push or just merge)

caiwei-ebay commented:
@cstamas As discussed in MRESOLVER-228, we will send multiple PRs following below sequence:

- DFS > BFS - preparation for parallel download, https://github.com/apache/maven-resolver/pull/144 is for this change.
- Skip & Reconcile - avoid unnecessary version resolution (MRESOLVER-228)
- Download descriptors in parallel (MRESOLVER-7)
